### DISCUSSION TITLE
Runtime Switchable Deployment Model Name

### DISCUSSION THREAD

**Discussion by mattahearn** [microsoft/chat-copilot#957] [Ideas]
Hello, I am attempting to make a toggle for the Azure OpenAI Deployment Name within User Settings so that a user can, at runtime, choose a different deployment i.e. gpt-35-turbo-deployment vs gpt4-deployment.  The issue I am running into is the Semantic Kernel/Provider are both created at app startup.  What would be the recommended approach to "update" the Configuration i.e. Deployment value for an existing dependency service?  Thank you for your help.
<img width="224" alt="image" src="https://github.com/microsoft/chat-copilot/assets/59938342/a7c20b63-e47a-4f3a-bb71-903b0bdeaed7">

The "Deployment" config resides within the AzureOpenAIText section of the appsettings.json file for the web api project.  
<img width="398" alt="image" src="https://github.com/microsoft/chat-copilot/assets/59938342/6054454e-724e-454d-9d86-2cfe79a39aa8">

**Comment by madsbolaris** [comment]
Within Semantic Kernel, you can author a custom service selector. I believe this can be used to dynamically alter which model is used during runtime. We have a sample here: https://github.com/microsoft/semantic-kernel/blob/9a4450622021ce003234863bcf4def9613ae1153/dotnet/samples/Concepts/Kernel/CustomAIServiceSelector.cs#L11

Do you think this addresses your challenges?

**Comment by aaronba** [comment]
Hey @matthewbolanos , Great sample. 

@mattahearn , you might want to consider mapping the sample concept approach for the CustomAIServiceSelector to the following code in the chat-copilot repo below:

https://github.com/microsoft/chat-copilot/blob/main/webapi/Services/SemanticKernelProvider.cs

**Comment by mattahearn** [comment]
[like]  Matt Ahearn reacted to your message:
________________________________

**Comment by mattahearn** [comment]
[like]  Matt Ahearn reacted to your message:
________________________________

**Comment by mattahearn** [comment]
@matthewbolanos your sample was helpful.  To extend this to the solution we need, how would you pass a parameter to your GptAIServiceSelector() from the Client in your example?  

I'm struggling to understand how I can send a runtime parameter from the Client to a DI service that is already added via  builder.Services.AddSingleton().  For example, the user on the client-side will chose "gpt4" or "gpt-35-turbo", then how do I pass this value down into the service collection so the Kernel uses the correct deployment i.e. "gpt4" or "gpt-35-turbo".

Thanks again for your help on this.

**Reply by sam0044** [reply]
@mattahearn did you end finding a solution for this?
